Windows: Enable ETW logging

jhowardmsft opened the pull request:
Signed-off-by: John Howard <jhoward@microsoft.com>

Enables ETW logging on Windows. (https://docs.microsoft.com/en-us/windows/desktop/etw/event-tracing-portal as a starting point for more information)

@kevpar @jterry75 FYI.

kevpar commented on cmd/containerd/command/main_windows.go:
```diff
@@ -89,3 +90,15 @@ func setupDumpStacks() {
+	// Hook isn't closed explicitly, as it will exist until process exit.
+	// The namespace sits under all other related container and Hyper-V providers.
+	// GUID is generated based on name - see Microsoft/go-winio/tools/etw-provider-gen.
+	hook, err := etwlogrus.NewHook("Microsoft.Virtualization.ContainerD")
```
The first part of the name generally indicates the company who owns the component, so we don't want to use Microsoft here. I'm not sure what's best for an open source project like this. Maybe ```Cncf.Containerd```?

jhowardmsft replied:
Ah, I see. Would simply "ContainerD" be sufficient? (I have the same issue with moby).

jterry75 replied:
I don't see an issue with ContainerD. I cant imagine this conflicts and the name to GUID will make this unique anyways.

kevpar replied:
Yeah, ContainerD is fine. I doubt anyone else will use it.

jhowardmsft replied:
OK, updated to `ContainerD` and GUID accordingly. I'll update my moby change to be `moby`.

jterry75 commented:
Nice! LGTM (fix the build break with the extra `return`)

jterry75 commented:
@jhowardmsft - Do we want to remove the event log logging hook?

kevpar commented:
@jterry75 don't think we should remove the event log hook yet. We can register a service to cause these ETW events to show up in event log, and then removing it's a good idea, but we should do that first in case anyone relies on it.

jterry75 commented:
@kevpar - I agree for Docker but this is new in CD anyways. Should we just effectively 'never add it'?

kevpar commented:
@jterry75 We might want to check with Nick first, I think event log -> fluentd might be their plan for logs.

jhowardmsft commented:
> Nice! LGTM (fix the build break with the extra return)

Removed. (Odd that lint didn't flag it on my dev box, but whatevs....)

jhowardmsft commented:
> @kevpar - I agree for Docker but this is new in CD anyways. Should we just effectively 'never add it'?

@jterry75 Let's take any action on this as a follow-up. Not related to this PR itself.

crosbymichael commented:
Looks like the vendor on linux is broken with this change.

```
These files were modified:
 M vendor/github.com/Microsoft/go-winio/internal/etw/etw.go
 M vendor/github.com/Microsoft/go-winio/internal/etw/eventdata.go
 M vendor/github.com/Microsoft/go-winio/internal/etw/eventdatadescriptor.go
 M vendor/github.com/Microsoft/go-winio/internal/etw/eventdescriptor.go
 M vendor/github.com/Microsoft/go-winio/internal/etw/eventmetadata.go
 M vendor/github.com/Microsoft/go-winio/internal/etw/eventopt.go
 M vendor/github.com/Microsoft/go-winio/internal/etw/fieldopt.go
 M vendor/github.com/Microsoft/go-winio/internal/etw/provider.go
 M vendor/github.com/Microsoft/go-winio/internal/etw/providerglobal.go
 M vendor/github.com/Microsoft/go-winio/internal/etw/ptr64_32.go
 M vendor/github.com/Microsoft/go-winio/internal/etw/ptr64_64.go
 M vendor/github.com/Microsoft/go-winio/pkg/etwlogrus/hook.go

The command "../project/script/validate/vendor" exited with 1.
```

jhowardmsft commented:
@crosbymichael Fixed now. Some Windows-style line endings had crept into go-winio. Updated that repo, cut a new release and updated the vendoring commit again.

crosbymichael commented on cmd/containerd/command/main_windows.go:
```diff
@@ -89,3 +90,12 @@ func setupDumpStacks() {
 	}()
 }
+
+func initETWLogging() {
```
can you just do this in an `init()` instead of calling it in main since we are not handling any of the errors outside of this function?

crosbymichael commented:
Overall, this looks good. I think we can move the initialization in an `init()` just to clean up the main.

jhowardmsft commented:
> Overall, this looks good. I think we can move the initialization in an init() just to clean up the main.

@crosbymichael Yes, makes sense. Updated.

crosbymichael commented:
LGTM
